### ISSUE TITLE
Add custom browser to Webapp manager

### ISSUE THREAD

**Issue by navidmafi** [linuxmint/webapp-manager#97]
I added custom browser to linux mint WebApp manager 
for example i added "Min Browser" to selections  
I hope this will help if anybody wants too!
 <br>
![ScreenShot 2021 03 16_17 18 24](https://user-images.githubusercontent.com/46371349/111319957-c3435000-867b-11eb-8517-22ccc359ec28.png)
 <br>
<h2>MAKE BACKUP BEFORE EDITING ANYTHING</h2>
1-make sure you can run the browser you want from a terminal command , for example i can run Min Browser with <code> min</code> command in terminal  so it's good   <br> 
2- go here : <code>/lib/webapp-manager/</code>  in your terminal  <br>
you can see <code>common.py</code> here - <strong>copy it to another location in case of problems</strong> <br>
 open it with root rights ( im using gedit: <code>sudo gedit common.py</code>)  <br>
3- Find this line : <code> browsers = []</code>  <br>
4- under this line you can add your browser with this format  <br>

<code> browsers.append(Browser(YOUR BROWSER TYPE , A NAME TO SHOW IN SELECTIONS, COMMAND THAT RUNS BROWSER, YOUR BROWSER BIN))</code>  <br>

<h4> Notes: </h4>
1- BROWSER TYPE is usually <code> BROWSER_TYPE_CHROMIUM</code>  for chromium based browsers <br>
2- usually the bin file location is /usr/bin/BROWSERNAME  <br>

<h2> Example: </h2>

`browsers.append(Browser(BROWSER_TYPE_CHROMIUM, "Min Browser", "min", "/usr/bin/min"))`

<br>
that was easy . i wanted to add my Whatsapp to webapps list but i wanted to run it from a super light browser not firefox <br>
i will update this post if any problems found <br>
I hope that helped! :smile:

**Comment by hsbasu** [comment]
It would be more user-friendly if this could be done via a GUI `choose custom browser` option. Then almost all the browsers could be accommodated.

**Comment by meiaihara06** [comment]
Yes, and I really want to add this feature. but some browsers failed to work properly with this method (could not pass URL argument properly i think?) . for example Min browser opened Whatsapp like this (If I remember correctly):
`httpswebwhatsappcom`

**Comment by vagabondsajib** [comment]
I have added `browsers.append(Browser(BROWSER_TYPE_EPIPHANY, "Tangram", "re.sonny.Tangram", "/usr/bin/re.sonny.Tangram"))` but upable to open any websites. If I try `web.shatsapp.com` it shows `Tangram —application-mode —profile="/home/mustafiz/.local/share/ice/epiphany/epiphany-Whatsapp6349" http://web.whatsapp.com`. I don't know what to do now! Any ideas?

**Comment by hsbasu** [comment]
> I have added `browsers.append(Browser(BROWSER_TYPE_EPIPHANY, "Tangram", "re.sonny.Tangram", "/usr/bin/re.sonny.Tangram"))` but upable to open any websites. If I try `web.shatsapp.com` it shows `Tangram —application-mode —profile="/home/mustafiz/.local/share/ice/epiphany/epiphany-Whatsapp6349" http://web.whatsapp.com`. I don't know what to do now! Any ideas?

This is an issue with Epiphany itself. Epiphany requires a file `.app` in the profile directory to be able to open a webapp in application mode. Also there are other issues like in Ubuntu 20.04 LTS it uses `epiphany-` to detect a webapp. However if you are using Ubuntu 21.04 or later (Probably it will be Ubuntu 20.10, I never checked though) the prefix to a profile directory is  `Org.Gnome.Epiphany.webapp-`(Never mind the upper and lower cases please). Not to mention `webapp-manager` does not copy the icon to the profile directory as `app-icon.png`. All or any of these issues might be affecting in your case.

**Comment by OatmanGeorge** [comment]
> I have added librewolf (flatpak):
> 
> `browsers.append(Browser(BROWSER_TYPE_FIREFOX_FLATPAK, "Librewolf (Flatpak)", "/var/lib/flatpak/exports/bin/io.gitlab.librewolf-community", "/var/lib/flatpak/exports/bin/io.gitlab.librewolf-community"))`
> 
> I added and tried to launch a web app using the new librewolf option, but I get an error message: "Your LibreWolf profile cannot be loaded. It may be missing or inaccessible"
> 
> Any idea what is wrong here?

Second that, I would also like to know how to completely and successfully add librewolf as a browser. After reading the thread and this here last comment, I was able to get past the error message by modifying the section in line 37 #Constants, adding a second line under 42 FIREFOX_FLATPAK_PROFILES_DIR = os.path.expanduser("~/.var/app/org.mozilla.firefox/data/ice/firefox"), replicating the same line but with FIREFOX_FLATPAK_PROFILES_DIR = os.path.expanduser("~/.var/app/io.gitlab.librewolf-community/data/"). 
But It's still wonky, opening different webapps with librewolf all show the same icon as the first opened one.
This issue is the same with Brave browser, btw, if you open a webapp with brave, all other subsequent webapps opened under Brave show the same icon as the first one, unless choosing to isolate profile when creating them.

**Comment by loyallucas** [comment]
Does anyone tried adding Opera? I tried but whenever I opened the webapp (facebook, for example) it just lanched a new opera window

**Comment by colour-in-space** [comment]
> > I have added librewolf (flatpak):
> > `browsers.append(Browser(BROWSER_TYPE_FIREFOX_FLATPAK, "Librewolf (Flatpak)", "/var/lib/flatpak/exports/bin/io.gitlab.librewolf-community", "/var/lib/flatpak/exports/bin/io.gitlab.librewolf-community"))`
> > I added and tried to launch a web app using the new librewolf option, but I get an error message: "Your LibreWolf profile cannot be loaded. It may be missing or inaccessible"
> > Any idea what is wrong here?
> 
> Second that, I would also like to know how to completely and successfully add librewolf as a browser. After reading the thread and this here last comment, I was able to get past the error message by modifying the section in line 37 #Constants, adding a second line under 42 `FIREFOX_FLATPAK_PROFILES_DIR = os.path.expanduser("~/.var/app/org.mozilla.firefox/data/ice/firefox")` , replicating the same line but with `FIREFOX_FLATPAK_PROFILES_DIR = os.path.expanduser("~/.var/app/io.gitlab.librewolf-community/data/")`. But It's still wonky, opening different webapps with librewolf all show the same icon as the first opened one. This issue is the same with Brave browser, btw, if you open a webapp with brave, all other subsequent webapps opened under Brave show the same icon as the first one, unless choosing to isolate profile when creating them.

Hi, librewolf is now in the ubuntu focal main repositories. I'm on KDE neon so that's great for me.
Adding it as a webapp now works, you still need to mainly add it as a new line in the common.py file, but this time not as flatpak.
Been working for months for me now, sorry I forgot to an update to my query.
